blog/user_roles.md

```diff
@@ -0,0 +1,294 @@
+    |> User.teacher_changeset(attrs)
+    |> Repo.insert()
+  end
+
+  @doc """
+  Creates a student.
+
+  ## Examples
+
+      iex> create_student(%{field: value})
+      {:ok, %User{}}
+
+      iex> create_student(%{field: bad_value})
+      {:error, %Ecto.Changeset{}}
+
+  """
+  def create_student(attrs \\ %{}) do
+    %User{}
+    |> User.student_changeset(attrs)
+    |> Repo.insert()
+  end
+```
+
+And all tests pass! I know what you're thinking. Why duplicate these functions based on only the role! But, yeah I like the explicitness for now and it could possibly diverge
+much more content-wise. So I'm gonna leave this as is. The tests are also very nice, concise and fast! Great!
+
+### CanCan or can it!
+
+Ok, you're still here... great! One more thing until I leave you. The whole idea of the introduction of roles in an app like this is that you want to restrict certain users
+from doing things they are not supposed to do. So called `authorization`! So in this scenario, most people start to google or hexl for a out-of-the-box solution for this
+and they'll find a handfull of them on hex indeed. But let's take a step back, read the excellent pow guides written by Dan, think again, and realize that Plug actually is all we need
+to implement this functionality ourselves, no dependencies added! I know there is a trend in Elixir land to get library maintainers going to get more tools in our toolbox, but
+the reason sometimes that these tools are not there is simply because we just don't need these dependencies because it is so straightforward to add ourselves. The big benefit in my
+opinion being that you fully understand what the code is doing, because you have written it all yourself! A win-win I'd say!
+
+Implementing a Plug is very straightforward and if you want to read up on this, make sure to visit the Plug docs, they're awesome! For now we'll use the suggested implementation
+that Dan provides in his guide and leave it at that for now. We'll definitely expand and refine it when we discover places in our app that need some more custom control, but for now I
+think I have drained your mental power enough for the day and leave with this Plug:
+
+```elixir
+defmodule StudentManagerWeb.Plugs.AuthorizationPlug do
+  @moduledoc """
+  This plug ensures that a user has a particular role.
+
+  ## Example
+
+      plug StudentManagerWeb.Plugs.AuthorizationPlug, [:student, :teacher]
+
+      plug StudentManagerWeb.Plugs.AuthorizationPlug, :teacher
+
+      plug StudentManagerWeb.Plugs.AuthorizationPlug, ~w(student teacher)a
+  """
+  alias StudentManagerWeb.Router.Helpers, as: Routes
+  alias Phoenix.Controller
+  alias Plug.Conn
+  alias Pow.Plug
+
+  @doc false
+  @spec init(any()) :: any()
+  def init(config), do: config
+
+  @doc false
+  @spec call(Conn.t(), atom()) :: Conn.t()
+  def call(conn, roles) do
+    conn
+    |> Plug.current_user()
+    |> has_role?(roles)
+    |> maybe_halt(conn)
+  end
+
+  defp has_role?(nil, _roles), do: false
```

Comment:
How about moving this to the `User` module? It's not specific to this plug.

Reply:
yes that is a possiblity if you are calling it on the user directly. but in this case it is possible to add the Plug to a pipeline and I think then it does make sense here, right? Or would it make more sense to define it in a user context and import that function here?